### PR TITLE
Feature/kotlin insight client sdk

### PR DESCRIPTION
**Pull request by HighKo**

### CODE REVIEW

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/error/InsightClientError.kt
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/error/InsightClientError.kt:1]: Maybe align to other messages? See `AtlasHttpClient`:
```Jira/Insight hat ein internes Problem festgestellt```

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/interfaces/GenericInsightObjectOperator.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * limitations under the License.
+ * #L%
+ */
+package com.linkedplanet.kotlininsightclient.api.interfaces
+
+import arrow.core.Either
+import com.linkedplanet.kotlininsightclient.api.error.InsightClientError
+import com.linkedplanet.kotlininsightclient.api.model.InsightObjectId
+
+interface GenericInsightObjectOperator<DomainType> {
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/interfaces/GenericInsightObjectOperator.kt:33]: Please add comments to public api

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightAttachmentOperator.kt [hunk 2]
```diff
@@ -79,18 +78,18 @@ class HttpInsightAttachmentOperator(private val context: HttpInsightClientContex
...
         context.httpClient.executeRestCall(
             "DELETE",
-            "/rest/insight/1.0/attachments/${attachmentId}",
+            "/rest/insight/1.0/attachments/${attachmentId.raw}",
             emptyMap(),
             null,
             "application/json"
         )
-            .map { it.body }
+            .map { }
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightAttachmentOperator.kt]: Add comment explaining empty map: `/*to Unit*/`

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightObjectOperator.kt [hunk 3]
```diff
@@ -27,161 +27,134 @@ import com.linkedplanet.kotlinhttpclient.api.http.GSON
...
+        func: suspend (InsightObject) -> Unit
     ): Either<InsightClientError, InsightObject> = either {
         val obj = createEmptyObject(objectTypeId)
         func(obj)
         val editItem = ObjectEditItem(
-            obj.objectTypeId,
+            obj.objectTypeId.raw,
             obj.getEditAttributes()
         )
         // TODO: ensure object type has the specified attributes
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightObjectOperator.kt:141]: Todo still relevant?

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightObjectOperator.kt [hunk 4]
```diff
@@ -243,13 +216,30 @@ class HttpInsightObjectOperator(private val context: HttpInsightClientContext) :
...
+                            ReferencedObject(InsightObjectId(ro.id), ro.label, ro.objectKey, ro.objectType?.let { ot ->
+                                ReferencedObjectType(InsightObjectTypeId(ot.id), ot.name)
+                            })
+                        },
+                        user = av.user?.run { InsightUser(displayName, name, emailAddress?: "", key) }
+                    )
+                }
             )
         }
         val objectSelf = "${context.baseUrl}/secure/insight/assets/${this.objectKey}"
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightObjectOperator.kt:239]: Ensure we have a test which tests that the self link works is callable (returns 200)

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/model/InsightObjectExtensions.kt
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/model/InsightObjectExtensions.kt:1]: We should ensure that those methods are tested properly, maybe with dedicated unit tests?

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/InsightObjectRestApiResponseObjects.kt [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    val name: String
+)
+
+internal data class ObjectTypeAttributeDefaultTypeApiResponse(
+    val id: Int,
+    val name: String
+)
+
+internal data class ObjectTypeSchemaAttributeReferenceTypeApiResponse(
+    val id: Int, // id 3, name = "Reference"; prefer an ENUM
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/InsightObjectRestApiResponseObjects.kt]: Change comment to point to `ReferenceKind` for more details

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/InsightObjectTypeRestApiResponseObjects.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    val id: Int,
+    val name: String,
+    val defaultType: ObjectTypeAttributeDefaultTypeApiResponse?,
+    val options: String,
+    val minimumCardinality: Int,
+    val maximumCardinality: Int,
+    val referenceType: ObjectTypeSchemaAttributeReferenceTypeApiResponse?,
+    val includeChildObjectTypes: Boolean,
+    val referenceObjectTypeId: Int?,
+    val type: Int // AttributeType 0=Default 1=Reference
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/InsightObjectTypeRestApiResponseObjects.kt]: Change comment to point to ReferenceKind for more details

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/InsightObjectTypeRestApiResponseObjects.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    // uniqueAttribute: Boolean
+    // hidden: Boolean
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/InsightObjectTypeRestApiResponseObjects.kt:48]: Remove

File: kotlin-insight-client/kotlin-insight-client-sdk/pom.xml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            <artifactId>insight-core-persistence</artifactId>
+            <version>${insight.version}</version>
+            <scope>provided</scope>
+        </dependency>
+<!--        <dependency>-->
+<!--            <groupId>com.atlassian.servicedesk</groupId>-->
+<!--            <artifactId>insight-core-services</artifactId>-->
+<!--            <version>${insight.version}</version>-->
+<!--            <scope>provided</scope>-->
+<!--        </dependency>-->
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/pom.xml]: Remove

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightAttachmentOperator.kt [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                val toByteArray = byteOutputStream.toByteArray()
+                toByteArray
+            }
+        }
+
+    override suspend fun uploadAttachment(
+        objectId: InsightObjectId,
+        filename: String,
+        byteArray: ByteArray
+    ): Either<InsightClientError, List<InsightAttachment>> =
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightAttachmentOperator.kt]: Why do we return a list of attachments?

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightAttachmentOperator.kt [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+    override suspend fun deleteAttachment(attachmentId: AttachmentId): Either<InsightClientError, Unit> =
+        catchAsInsightClientError {
+            objectFacade.deleteAttachmentBean(attachmentId.raw).toString()
+        }
+
+    private fun beanToInsightAttachment(bean: AttachmentBean): InsightAttachment = bean.run {
+        val url = attachmentUrlResolver.buildUrlForAttachment(bean)
+        val humanReadableFileSize = humanReadableByteCountSI(fileSize)
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightAttachmentOperator.kt:113]: Consider using `org.apache.commons.io.FileUtils.byteCountToDisplaySize(fileSize)`

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightObjectOperator.kt [hunk 11]
```diff
@@ -27,161 +27,134 @@ import com.linkedplanet.kotlinhttpclient.api.http.GSON
...
-            pageFrom,
-            pageTo,
-            perPage
-        ).bind()
+        getObjectsByIQL(fullIql, pageIndex, pageSize).bind()
     }
 
     override suspend fun getObjectsByIQL(
         iql: String,
-        pageFrom: Int,
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightObjectOperator.kt:75]: We should add the possibility to retrieve all objects for an iql without pagination. Let's discuss this one.

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+        }
+
+    override suspend fun createObject(
+        objectTypeId: InsightObjectTypeId,
+        func: suspend (InsightObject) -> Unit // to configure the model
+    ): Either<InsightClientError, InsightObject> =
+        catchAsInsightClientError {
+            val freshInsightObject = createEmptyDomainObject(objectTypeId)
+            func(freshInsightObject)
+            val objectTypeBean = objectTypeFacade.loadObjectType(objectTypeId.raw)
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt:154]: `objectTypeFacade.loadObjectType` is called twice, here and in `createEmptyDomainObject`, maybe we can pass the `objectTypeBean` on to the latter method.

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+    override suspend fun updateObject(obj: InsightObject): Either<InsightClientError, InsightObject> =
+        catchAsInsightClientError {
+            val objectBean = objectFacade.loadObjectBean(obj.id.value).createMutable()
+            setAttributesForObjectBean(obj, objectBean)
+            objectBean.objectTypeId = obj.objectTypeId.raw
+            objectBean.objectKey = obj.objectKey
+            objectFacade.storeObjectBean(objectBean)
+        }.flatMap { it.toInsightObject() }
+
+    private fun setAttributesForObjectBean(
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt]: Please move down in private section

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+        objectBean: ObjectBean,
+        objectTypeBean: ObjectTypeBean,
+        objectTypeAttributeBeans: List<ObjectTypeAttributeBean>,
+        hasAttachments: Boolean
+    ): Either<InsightClientError, InsightObject> = either {
+        val attributes = objectBean.objectAttributeBeans.map { objAttributeBean ->
+            val objTypeAttributeBean = objectTypeAttributeBeans.typeForBean(objAttributeBean).bind()
+            createInsightAttribute(objAttributeBean, objTypeAttributeBean).bind()
+        }
+        val objectSelf = "${baseUrl}/secure/insight/assets/${objectBean.objectKey}"
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt:229]: Is there no `self` attribute?

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+            val hasAttachments = objectFacade.findAttachmentBeans(id).isNotEmpty()
+            return@toInsightObject createInsightObject(
+                this@toInsightObject,
+                objectType,
+                objectTypeAttributeBeans,
+                hasAttachments
+            )
+        }
+
+    private suspend fun createInsightObject(
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt]: maybe call this one `mapBeansToInsightObject`

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    private fun List<ObjectTypeAttributeBean>.typeForBean(
+        objAttributeBean: ObjectAttributeBean
+    ): Either<InsightClientError, ObjectTypeAttributeBean> =
+        singleOrNull { it.id == objAttributeBean.objectTypeAttributeId }
+            ?.right()
+            ?: ObjectTypeNotFoundError().left()
+
+    private suspend fun createInsightAttribute(
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectOperator.kt]: maybe call this one `mapBeansToInsigtAttribute`

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/HttpEditModel.kt [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+internal fun InsightObject.toEditObjectItem() =
+    ObjectEditItem(
+        objectTypeId.raw,
+        getEditAttributes()
+    )
+
+internal fun InsightObject.getEditAttributes(): List<ObjectEditItemAttribute> =
+    this.attributes.map { insightAttr ->
+        val values = insightAttr.value.map {
+            if (insightAttr.attributeType == InsightObjectAttributeType.REFERENCE) {
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/HttpEditModel.kt:40]: Mapping for `InsightObjectAttributeType.USER` missing

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectTypeOperator.kt [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+            ObjectTypeAttributeBean.Type.REFERENCED_OBJECT -> InsightObjectAttributeType.REFERENCE
+            ObjectTypeAttributeBean.Type.USER -> InsightObjectAttributeType.USER
+            ObjectTypeAttributeBean.Type.CONFLUENCE -> InsightObjectAttributeType.CONFLUENCE
+            ObjectTypeAttributeBean.Type.GROUP -> InsightObjectAttributeType.GROUP
+            ObjectTypeAttributeBean.Type.VERSION -> InsightObjectAttributeType.VERSION
+            ObjectTypeAttributeBean.Type.PROJECT -> InsightObjectAttributeType.PROJECT
+            ObjectTypeAttributeBean.Type.STATUS -> InsightObjectAttributeType.STATUS
+        }
+
+    private fun mapDefaultType(defaultType: ObjectTypeAttributeBean.DefaultType) =
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/SdkInsightObjectTypeOperator.kt:105]: Please add return type explicitly

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/services/ReverseEngineeredAttachmentUrlResolver.kt [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+ */
+package com.linkedplanet.kotlininsightclient.sdk.services
+
+import com.atlassian.jira.component.ComponentAccessor.getComponent
+import com.atlassian.jira.config.properties.ApplicationProperties
+import com.riadalabs.jira.plugins.insight.services.model.AttachmentBean
+import java.util.regex.Pattern
+
+
+internal class ReverseEngineeredAttachmentUrlResolver {
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/services/ReverseEngineeredAttachmentUrlResolver.kt:33]: Just add a small comment why that's necessary and where the sources can be found

File: kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/services/ReverseEngineeredFileManager.kt [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+import com.atlassian.jira.component.ComponentAccessor.getComponent
+import com.atlassian.jira.config.util.AttachmentPathManager
+import com.atlassian.jira.issue.AttachmentManager
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileNotFoundException
+import java.io.InputStream
+
+internal class ReverseEngineeredFileManager{
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-sdk/src/main/kotlin/com/linkedplanet/kotlininsightclient/sdk/services/ReverseEngineeredFileManager.kt:36]: Just add a small comment why that's necessary and where the sources can be found

File: kotlin-insight-client/kotlin-insight-client-test-applink/pom.xml [hunk 21]
```diff
@@ -93,7 +91,7 @@
         <dependency>
             <groupId>com.atlassian.plugins</groupId>
             <artifactId>atlassian-plugins-osgi-testrunner</artifactId>
-            <version>2.0.1</version>
+            <version>2.0.3</version>
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-test-applink/pom.xml]: Extract that version and all its' occurrences as var into the root project

File: kotlin-insight-client/kotlin-insight-client-test-applink/src/test/kotlin/it/InsightApplinkClientTest.kt [hunk 22]
```diff
@@ -58,6 +59,9 @@ class InsightApplinkClientTest constructor(
 
         clientContext = HttpInsightClientContext("http://localhost:8080", httpClient)
 
+        GenericInsightObjectOperatorImpl.insightObjectOperator = insightObjectOperator
+        GenericInsightObjectOperatorImpl.insightObjectTypeOperator = insightObjectTypeOperator
+        GenericInsightObjectOperatorImpl.insightSchemaOperator = insightSchemaOperator
         println("### Starting InsightApplinkClientTest")
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-test-applink/src/test/kotlin/it/InsightApplinkClientTest.kt:65]: Started InsightApplinkClientTest - please align this logs in all test modules

File: kotlin-insight-client/kotlin-insight-client-test-base/src/main/kotlin/com/linkedplanet/kotlininsightclient/InsightAttachmentOperatorTest.kt [hunk 23]
```diff
@@ -19,40 +19,161 @@
...
+import java.util.zip.ZipInputStream
 
 interface InsightAttachmentOperatorTest {
     val insightObjectOperator: InsightObjectOperator
     val insightAttachmentOperator: InsightAttachmentOperator
 
     @Test
-    fun testAttachments() {
+    fun attachmentTestGetAttachments() = runBlocking {
         println("### START attachment_testGetAndDownloadAttachments")
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-test-base/src/main/kotlin/com/linkedplanet/kotlininsightclient/InsightAttachmentOperatorTest.kt:46]: Correct output - Check logs for all methods!

Is there an automated approach to print such logs?

File: kotlin-insight-client/kotlin-insight-client-test-base/src/main/kotlin/com/linkedplanet/kotlininsightclient/InsightHistoryOperatorTest.kt [hunk 24]
```diff
@@ -35,8 +38,10 @@ interface InsightHistoryOperatorTest {
 
         runBlocking {
             val country = insightObjectOperator.getObjectByName(InsightObjectType.Country.id, "Germany").orNull()!!
-            val historyItems = insightHistoryOperator.getHistory(country.id).orNull()!!
-            assertTrue(historyItems.isNotEmpty())
+            val history = insightHistoryOperator.getHistory(country.id).orNull()!!
+            assertTrue(history.historyItems.isNotEmpty())
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-test-base/src/main/kotlin/com/linkedplanet/kotlininsightclient/InsightHistoryOperatorTest.kt:42]: We should not mix JUnit and Hamcrest assertions

File: kotlin-insight-client/kotlin-insight-client-test-base/src/main/kotlin/com/linkedplanet/kotlininsightclient/InsightSchemaOperatorTest.kt [hunk 25]
```diff
@@ -30,11 +32,14 @@ interface InsightSchemaOperatorTest {
     @Test
     fun testSchemaLoad() {
         println("### START schema_testSchemaLoad")
+        val schemas = runBlocking { insightSchemaOperator.getSchemas() }.orFail()
+        val retrievedSchema = schemas.firstOrNull { it.id == SchemaObject.ITest.id }!!
+        assertThat(retrievedSchema.name, equalTo("ITest"))
+        assertThat(retrievedSchema.objectTypeCount,  greaterThanOrEqualTo (7)) // 7 when test was written
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-test-base/src/main/kotlin/com/linkedplanet/kotlininsightclient/InsightSchemaOperatorTest.kt:38]: Please extract those IDs into the model

File: kotlin-insight-client/kotlin-insight-client-test-sdk/pom.xml [hunk 26]
```diff
@@ -0,0 +1,301 @@
...
+
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-antrun-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>show-home-warning</id>
+                        <configuration>
+                            <target>
+                                <input message="ATTENTION!${line.separator}Remove the license information which is stored in the database. Therefore stop Jira, connect to the database (e.g. with IntelliJ) and finally remove the entries from the license tables.${line.separator}Press enter to continue..."/>
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-test-sdk/pom.xml:261]: Ensure your test-resources don't contain license information

File: kotlin-insight-client/kotlin-insight-client-test-sdk/src/main/kotlin/com/linkedplanet/plugin/jira/insightclient/test/api/PluginComponent.kt [hunk 27]
```diff
@@ -17,9 +17,10 @@
  * limitations under the License.
  * #L%
  */
-package com.linkedplanet.kotlinhttpclient.error
+package com.linkedplanet.plugin.jira.insightclient.test.api
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-test-sdk/src/main/kotlin/com/linkedplanet/plugin/jira/insightclient/test/api/PluginComponent.kt:20]: incorrect package

File: kotlin-jira-client/kotlin-jira-client-test-base/src/main/kotlin/com/linkedplanet/kotlinjiraclient/util/JiraUserTestHelper.kt [hunk 28]
```diff
@@ -35,12 +38,12 @@ class JiraUserTestHelper {
...
-            assertEquals(expectedDisplayName, jiraUser.displayName)
-            assertEquals(expectedEmail, jiraUser.emailAddress)
+            assertThat(jiraUser, notNullValue())
+            assertThat(jiraUser!!.name, equalTo(expectedName))
+            assertThat(jiraUser.displayName, equalTo(expectedDisplayName))
+            assertThat(jiraUser.emailAddress, equalTo(expectedEmail))
             assertTrue(jiraUser.key.isNotEmpty())
-            assertNotNull(jiraUser.avatarUrl)
+            assertThat(jiraUser.avatarUrl, notNullValue())
             assertTrue(jiraUser.avatarUrl!!.startsWith("https://www.gravatar.com/avatar"))
```
Comment (sjahreis) [inline kotlin-jira-client/kotlin-jira-client-test-base/src/main/kotlin/com/linkedplanet/kotlinjiraclient/util/JiraUserTestHelper.kt:47]: Mixture of Hamcrest and JUnit assertions

File: kotlin-http-client/kotlin-http-client-ktor/src/main/kotlin/com/linkedplanet/kotlinhttpclient/ktor/KtorHttpClient.kt [hunk 29]
```diff
@@ -65,9 +63,7 @@ class KtorHttpClient(
         requestBuilder.url("$baseUrl/$path$parameterString")
         requestBuilder.contentType(parsedContentType)
         if (bodyIn != null) {
-            // Keep JsonParser instantiation for downwards compatibility
-            @Suppress("DEPRECATION")
-            requestBuilder.body = JsonParser().parse(bodyIn)
+            requestBuilder.setBody(bodyIn)
```
Comment (sjahreis) [inline kotlin-http-client/kotlin-http-client-ktor/src/main/kotlin/com/linkedplanet/kotlinhttpclient/ktor/KtorHttpClient.kt:66]: Please help, what was the reason for the "old" comment? I can't remember...

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt [hunk 30]
```diff
@@ -0,0 +1,233 @@
...
+            val value = prop.get(domainObject)
+            when (attribute.type) {
+                InsightObjectAttributeType.DEFAULT -> {
+                    if (value is List<Any?>) {
+                        insightObject.setValueList(attribute.id, value)
+                    } else {
+                        insightObject.setValue(attribute.id, value)
+                    }
+                }
+                InsightObjectAttributeType.REFERENCE -> {
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt:121]: Add USER case?

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt [hunk 31]
```diff
@@ -0,0 +1,233 @@
...
+@Experimental
+class GenericInsightObjectOperatorImpl<DomainType : Any>(
+    private val klass: KClass<DomainType>,
+    // having InsightObject as return type will prevent us to change the internal implementation later on
+    private val insightObjectForDomainObject: suspend (objectTypeId: InsightObjectTypeId, domainObject: DomainType) -> Either<InsightClientError, InsightObject?>,
+    private val referenceAttributeToValue: suspend (attribute: InsightAttribute) -> Any? = { null },
+    private val attributeToReferencedObjectId: suspend (attribute: ObjectTypeSchemaAttribute, Any?) -> List<InsightObjectId> = { _, _ -> emptyList() },
+) : GenericInsightObjectOperator<DomainType> {
+    private val props: Collection<KProperty1<DomainType, *>> = klass.memberProperties
+    var objectTypeSchema: ObjectTypeSchema
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt]: private?

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt [hunk 32]
```diff
@@ -0,0 +1,233 @@
...
+
+    init {
+        runBlocking {
+            objectTypeSchema = objectTypeSchemaFromKClass().orNull()!!
+            attrsMap = objectTypeSchema.attributes.associateBy { it.name.lowercase() }
+        }
+    }
+
+    override suspend fun delete(domainObject: DomainType): Either<InsightClientError, Unit> = either {
+        val insightObject = insightObjectForDomainObject(objectTypeSchema.id, domainObject).bind() ?: return@either
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt:91]: why do we request the object before deletion? isn't that obsolete?

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt [hunk 33]
```diff
@@ -0,0 +1,233 @@
...
+    // having InsightObject as return type will prevent us to change the internal implementation later on
+    private val insightObjectForDomainObject: suspend (objectTypeId: InsightObjectTypeId, domainObject: DomainType) -> Either<InsightClientError, InsightObject?>,
+    private val referenceAttributeToValue: suspend (attribute: InsightAttribute) -> Any? = { null },
+    private val attributeToReferencedObjectId: suspend (attribute: ObjectTypeSchemaAttribute, Any?) -> List<InsightObjectId> = { _, _ -> emptyList() },
+) : GenericInsightObjectOperator<DomainType> {
+    private val props: Collection<KProperty1<DomainType, *>> = klass.memberProperties
+    var objectTypeSchema: ObjectTypeSchema
+    private var attrsMap: Map<String, ObjectTypeSchemaAttribute>
+
+    companion object {
```
Comment (sjahreis) [inline kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt:75]: We should change those to object attributes

**Review by sjahreis — CHANGES_REQUESTED** [review]

**Comment by sjahreis** [comment]
- [x] Add a README entry for the kotlin-insight-client and its' submodules
- [x] kotlin-jira-client README: ausbessern "kotlin-jira-client-test-ktor: Tests the kotlin-jira-client-sdk in a Jira Plugin"
- [x] kotlin-jira-client README: Badges entfernen

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt (hunk 30 above)
Comment (HighKo) [reply kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt:121]: Its still experimental. Lets not add more work right now.

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt (hunk 33 above)
Comment (HighKo) [reply kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt:75]: I agree, there is a comment for that already inside the class.

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt (hunk 31 above)
Comment (HighKo) [reply kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/experimental/GenericInsightObjectOperatorImpl.kt]: this is public, so clients could use it to add missing functionality

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightObjectOperator.kt (hunk 4 above)
Comment (HighKo) [reply kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/HttpInsightObjectOperator.kt:239]: callable reicht nicht, weil bei der login Seite auch 200 zurück kommt :D

File: kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/model/InsightObjectExtensions.kt
Comment (HighKo) [reply kotlin-insight-client/kotlin-insight-client-api/src/main/kotlin/com/linkedplanet/kotlininsightclient/api/model/InsightObjectExtensions.kt:1]: habs mir aufgeschrieben für anderen branch

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/InsightObjectTypeRestApiResponseObjects.kt (hunk 7 above)
Comment (HighKo) [reply kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/InsightObjectTypeRestApiResponseObjects.kt:48]: ist schon gut, dass man sieht, dass da noch mehr zurück kommt

File: kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/HttpEditModel.kt (hunk 17 above)
Comment (HighKo) [reply kotlin-insight-client/kotlin-insight-client-http/src/main/kotlin/com/linkedplanet/kotlininsightclient/http/model/HttpEditModel.kt:40]: different branch

**Pull request merged**